test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -0,0 +1,641 @@
+      });
+      it("should add the number of users correctly", async () => {
+        await pageAction._handleClick();
+        const footerUsers = elements["cfr-notification-footer-users"];
+        assert.isNull(footerUsers.getAttribute("hidden"));
+        assert.equal(footerUsers.getAttribute("value"), `${fakeRecommendation.content.addon.users} users`);
+      });
+      it("should send the right telemetry", async () => {
+        await pageAction._handleClick();
+        assert.calledWith(dispatchStub, {
+          type: "DOORHANGER_TELEMETRY",
+          data: {
+            useClientID: true,
+            action: "cfr_user_event",
+            source: "CFR",
+            message_id: fakeRecommendation.id,
+            bucket_id: fakeRecommendation.content.bucket_id,
+            event: "CLICK_DOORHANGER"
+          }
+        });
+      });
+      it("should call PopupNotifications.show with the right arguments", async () => {
+        await pageAction._handleClick();
+        assert.calledWith(
+          global.PopupNotifications.show,
+          fakeBrowser,
+          "contextual-feature-recommendation",
+          fakeRecommendation.content.addon.title,
+          "cfr",
+          sinon.match.any,
+          sinon.match.any,
+          {
+            popupIconURL: fakeRecommendation.content.addon.icon,
+            hideClose: true,
+            eventCallback: pageAction._popupStateChange
+          }
+        );
+
+        const [, , , , mainAction, [secondaryAction]] = global.PopupNotifications.show.firstCall.args;
+
+        // Is the mainAction correct?
+        assert.deepEqual(mainAction.label, {value: "Primary Button", attributes: {accesskey: "p"}});
+        sandbox.spy(pageAction, "hide");
+        mainAction.callback();
+        assert.calledOnce(pageAction.hide);
+        // Should block the message
+        assert.calledWith(dispatchStub, {
+          type: "BLOCK_MESSAGE_BY_ID",
+          data: {id: fakeRecommendation.id}
+        });
+        // Should trigger the action
+        assert.calledWith(
+          dispatchStub,
+          {type: "USER_ACTION", data: {id: "primary_action"}},
+          {browser: fakeBrowser}
+        );
+        // Should send telemetry
+        assert.calledWith(dispatchStub, {
+          type: "DOORHANGER_TELEMETRY",
+          data: {
+            useClientID: true,
+            action: "cfr_user_event",
+            source: "CFR",
+            message_id: fakeRecommendation.id,
+            bucket_id: fakeRecommendation.content.bucket_id,
+            event: "INSTALL"
+          }
+        });
+        // Should remove the recommendation
+        assert.isFalse(CFRPageActions.RecommendationMap.has(fakeBrowser));
+
+        // Is the secondaryAction correct?
+        assert.deepEqual(secondaryAction.label, {value: "Secondary Button", attributes: {accesskey: "s"}});
+        sandbox.resetHistory(); // .hide was called above so reset the call count
```

Comment:
I would also split this up and use a `beforeEach` block if setup is too complicated/long